lib/foreman/renderer/scope/macros/host_template.rb

```diff
@@ -121,6 +121,51 @@ def ks_console
             (@port && @baud) ? "console=ttyS#{@port},#{@baud}" : ''
           end
 
+          apipie :method, 'Generates commands for package(s) installation' do
+            description 'Supported OS families: Redhat, Debian & Suse'
+            required :packages, String, desc: "Package(s) to install"
+            returns String, desc: 'Installation commands'
+            example "CentOS host: install_packages('pkg1') #=> yum -y install pkg1"
+            example "Fedora host: install_packages('pkg1 pkg2') #=> dnf -y install pkg1 pkg2"
+          end
+          def install_packages(packages)
+            return '' if packages.blank?
+
+            banner = <<~CMD
+              echo '#'
+              echo '# Installing packages'
+              echo '#'
+            CMD
+
+            case host.operatingsystem&.family
+            when 'Redhat'
+              <<~CMD
+                #{banner}
+                if [ -f /usr/bin/dnf ]; then
```

Comment:
I still think we should use host OS properties (from our models) to determine this rather than at runtime. From our Jenkins nodes running RHEL 7:
```console
$ rpm -qf /usr/bin/dnf
dnf-4.0.9.2-2.el7_9.noarch
$ cat /etc/os-release 
NAME="Red Hat Enterprise Linux Server"
VERSION="7.9 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.9"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.9 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.9:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.9
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.9"
```

However, we generally don't use dnf to install things. It's there to allow some RPM actions in our packaging workflows.

(Yes, this is all very messy)

Reply:
Yeah but there is so many cases that this is the easiest solution to catch all of them. Simple and effective.

> However, we generally don't use dnf to install things. 

But in previous reviews you wanted to use `dnf`. Why is it bad now?

Reply:
DNF on EL8 is good, DNF on EL7 is possible, but not what I would expect to be used. That's my concern. So what was mentioned in https://github.com/theforeman/foreman/pull/8298#discussion_r609032259 sounds close to what I would expect. I would expect (written in a shortened form):
```
if (is_fedora && version >= 22) || (!is_fedora && version >= 8)
```

Reply:
But the problem here are the future clones of CentOS (Rocky & others), where the condition `(!is_fedora && version >= 8)` can be in future true but they won't have `dnf`, just `yum`. 

Anyway, here is a table with the options, is this correct?

OS | version | package manager
------------ | -------------| -------------
Fedora | all versions | `dnf`
CentOS |all|`yum`
RHEL|<= 7| `yum`
RHEL|>= 8|`dnf`
CentOS clones|?|`yum`

Reply:
No, that's not correct. I've been too long in the OS world that I have to stop and remember this isn't just common knowledge to everyone. Here's the background.

Fedora started to use DNF in version 22, prior to that it was yum.

CentOS Linux is a rebuild of RHEL. For almost all intents and purposes, you can treat it as the same OS. That also means CentOS Linux is by definition behind RHEL (or at most, in sync). CentOS Linux has a bug-for-bug compatible with RHEL policy. This means that if a bug is reported and it's also present in RHEL, it's considered a WONTFIX and they tell you to get it fixed in RHEL instead. Fixing the bug would introduce different behavior from RHEL and is considered a bug. Note that CentOS Linux 8 will be EOL at the end of the year while 7 will remain supported until 2024 when regular RHEL support ceases.

Then we have the new kid CentOS Stream which is being positioned as an upstream of RHEL. This means it's similar, but it can actually be ahead of RHEL. Another big difference is that CentOS Stream does **not** have a minor version. There is just 8, not 8.x. This also means we will see a CentOS Stream 9 before RHEL 9.

Then there are other rebuilds of RHEL (not CentOS) like Rocky and AlmaLinux. They, like CentOS Linux, aim to be pure RHEL rebuilds. This is why you see the concept of OS Family: we can treat the Red Hat OS family (with the exception of Fedora) the same. This is also why you see EL used: it's RHEL and all its rebuilds.

Because of this, you see patterns in Puppet like
```puppet
case $facts['os']['family'] {
  'RedHat': {
    case $facts['os']['name'] {
      'Fedora': {
        # Fedora specific code
      }
      default: {
        if versioncmp($facts['os']['release']['major'], '8') >= {
          # EL8+
        } else {
          # EL7 and older
        }
      }
    }
  }
  'Debian': {
    case $facts['os']['name'] {
      'Ubuntu': {}
      default: {}
    }
  }
}
```

Foreman has its roots in Puppet and imported facts so it copied the OS family concept. That makes you understand the OS model in our database.

Then there are RHEL derivatives like OracleLinux and AmazonLinux. They are based on RHEL, but unlike rebuilds they aren't quite the same. Some base packages have changed or may not be shipped. OracleLinux is still quite close to RHEL while AmazonLinux has made much bigger changes. Facter no longer considers AmazonLinux part of the Red Hat OS family because of that while OracleLinux is. Due to all of this we consider them best effort supported. It is up to users of those distros to send fixes if there are problems and most developers never touch them. 

So in conclusion, it becomes:

OS | version | package manager
------------ | -------------| -------------
Fedora | >= 22 | `dnf`
Fedora | < 22 | `yum`
EL|>= 8|`dnf`
EL|< 8| `yum`

Does that help?

Reply:
> Does that help?

Yes, a lot, thanks :) I'll update the PR according to your table.